File: qts-master/qts-admin/src/main/java/org/qts/admin/core/AcctInst.java
package org.qts.admin.core;

import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import org.qts.admin.entity.AcctDesc;
import org.qts.admin.exception.BizException;
import org.qts.common.entity.Enums;
import org.qts.common.entity.Message;
import org.qts.common.entity.acct.AcctDetail;
import org.qts.common.entity.acct.AcctInfo;
import org.qts.common.entity.config.AcctConf;
import org.qts.common.entity.event.MessageEvent;
import org.qts.common.entity.trade.Order;
import org.qts.common.entity.trade.Position;
import org.qts.common.entity.trade.Tick;
import org.qts.common.entity.trade.Trade;
import org.qts.common.rpc.tcp.client.MsgHandler;
import org.qts.common.rpc.tcp.client.TcpClient;
import org.qts.common.utils.ProcessUtil;
import org.qts.common.utils.SequenceUtil;
import org.qts.common.utils.SpringUtils;
import org.springframework.beans.BeanUtils;

import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

@Data
@Slf4j
public class AcctInst  implements MsgHandler {
    private AcctDetail acct;
    private String id;
    private Enums.ACCT_STATUS status;
    private Integer pid = 0;
    private String updateTimes;

    private TcpClient tcpClient;

    public AcctInst(AcctConf conf){
        this.id = conf.getId();
        this.acct = new AcctDetail(conf);
        this.status = Enums.ACCT_STATUS.UNSTARTED;
        this.updateTimes = LocalTime.now().format(DateTimeFormatter.ofPattern("HHmmss"));

        this.tcpClient = new TcpClient(this.acct.getId(),this.acct.getConf().getPort(),this);
    }
    public AcctDesc getAcctInstDesc(){
        AcctDesc acctDesc =new AcctDesc();
        BeanUtils.copyProperties(this.acct,acctDesc);
        acctDesc.setStatus(this.status);
        acctDesc.setUpdateTimes(this.updateTimes);
        return acctDesc;
    }

    public void chekcInstStatus(){
        //检查账户进程是否启动
        int pid= ProcessUtil.getProcess("qts-trader","acctId="+this.getId());
        if(pid<=0){
            if(this.pid>0)
                log.info("账户[{}]已停止,pid:{}",this.getId(),this.pid);
            this.setStatus(Enums.ACCT_STATUS.UNSTARTED);
            this.getTcpClient().stop();
            this.pid=0;
            return;
        }
        if(this.pid==0)
            log.info("账户[{}]已启动,pid:{}",this.getId(),pid);
        this.pid = pid;
        this.setStatus(Enums.ACCT_STATUS.CONNING);
        //启动客户端
        this.tcpClient.start();
        if(this.tcpClient.isConnected())
            this.setStatus(Enums.ACCT_STATUS.READY);
       this.updateTimes = LocalTime.now().format(DateTimeFormatter.ofPattern("HHmmss"));

       //广播状态
        SpringUtils.pushEvent(new MessageEvent(new Message(Enums.MSG_TYPE.ON_ACCT,this.getAcctInstDesc())));

    }

    public void startProcess(){
        try {
            List<String> cmds = new ArrayList<String>();
            cmds.add("java");
            cmds.add("-DacctId="+this.id);
            cmds.add("-jar");
            //cmds.add("-Xdebug -Xnoagent -Djava.compiler=NONE -Xrunjdwp:transport=dt_socket,server=y,suspend=y,address=5005");
            cmds.add("qts-trader/target/qts-trader.jar");
            pid=ProcessUtil.startProcess(cmds)  ;
            this.pid=pid;
        }catch (Exception ex){
            log.error("启动账户[{}]失败",this.id,ex);
        }
    }

    public Message request(Message req){
        String requestId = SequenceUtil.getLocalSerialNo(16);
        req.setRequestId(requestId);
        log.info("request==>req:{}",req);
        if(this.status!= Enums.ACCT_STATUS.READY && !tcpClient.isConnected()){
            throw  new BizException("账户未就绪");
        }
        Message rsp= tcpClient.request(req);
        log.info("request==>rsp:{}",rsp);
        return rsp;
    }

    @Override
    public void onMessage(Message msg) {
        switch (msg.getType()){
            case ON_DETAIL -> {
                AcctDetail acctDetail = msg.getData(AcctDetail.class);
                if(acctDetail!=null){
                    this.acct =acctDetail;
                    log.info("更新账户明细,{}",this.acct.getId());
                }

            }
            case ON_ACCT -> {
                AcctInfo acctInfo = msg.getData(AcctInfo.class);
                BeanUtils.copyProperties(this.acct,acctInfo);
                log.info("更新账户信息,{}",this.acct.getId());
            }
            case ON_POSITION -> {
                Position pos = msg.getData(Position.class);
                this.acct.getPositions().put(pos.getId(),pos);
            }
            case ON_TRADE -> {
                Trade trade = msg.getData(Trade.class);
                this.acct.getTradeList().put(trade.getTradeID(),trade);
            }
            case ON_ORDER -> {
                Order order = msg.getData(Order.class);
                this.acct.getOrders().put(order.getOrderSysID(),order);
                if(order.isFinished())
                    this.acct.getOrders().remove(order.getOrderSysID());
            }
            case ON_TICK -> {
                Tick tick = msg.getData(Tick.class);
                this.acct.getTicks().put(tick.getSymbol(),tick);
            }
        }
    }
}

File: qts-admin/core/acct/acct_inst.py
from core.rpc import RpcClient, RpcHandler
from core.rpc import MsgType
from model.acct import  AcctConf,AcctInfo
import subprocess
import os
import psutil




class AcctInst(object):
    def __init__(self, config):
        self.config: AcctConf = config
        self.acct_id = config.id
        self.acct_client: RpcClient = None
        acct_info = AcctInfo(id=config.id,group=config.group,name=config.name,enable=config.enable,conf=config)
        self.acct_info: AcctInfo = acct_info
        self.acct_detail = None


    def start_inst(self):     
        #连接账户进程
        req_address = self.config.req_address
        pub_address = self.config.pub_address
        self.acct_client = RpcClient()
        self.acct_client.register_handler(self.create_handler())
        self.acct_client.start(req_address, pub_address)

    


    def stop_inst(self):
        self.acct_client.stop()
        self.acct_client = None

    def connect(self):
        self.acct_client.req(MsgType.CONNECT)

    def disconnect(self):
        self.acct_client.req(MsgType.DISCONNECT)

    def close(self):
        self.acct_client.req(MsgType.CLOSE)

    def get_positions(self):
        self.acct_client.req(MsgType.GET_POSITIONS)
        pass

    def get_trades(self):
        self.acct_client.req(MsgType.GET_TRADES)
        pass

    def get_acct_info(self):
        self.acct_client.req(MsgType.GET_ACCT_INFO)
        pass

    def get_acct_detail(self):
        self.acct_client.req(MsgType.GET_ACCT_DETAIL)
        pass

    def send_order(self, order):
        self.acct_client.req(MsgType.SEND_ORDER, order)
        pass

    def cancel_order(self, orderid):
        self.acct_client.req(MsgType.CANCEL_ORDER, orderid)
        pass


    def create_handler(self):
        topic_handler = RpcHandler()

        @topic_handler.register_handler(MsgType.ON_TICK)
        def on_tick(data):
            config = self.config
            pass

        @topic_handler.register_handler(MsgType.ON_ORDER)
        def on_order(data):
            pass

        @topic_handler.register_handler(MsgType.ON_POSITION)
        def on_position(data):
            pass

        @topic_handler.register_handler(MsgType.ON_LOG)
        def on_log(data):
            pass

        return topic_handler
